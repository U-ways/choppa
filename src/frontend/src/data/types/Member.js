export default class Member {
  constructor(id, name, chapter) {
    this._id = id;
    this._name = name;
    this._chapter = chapter;
  }

  get id() {
    return this._id;
  }

  get name() {
    return this._name;
  }

  set name(newName) {
    this._name = newName;
  }

  get chapter() {
    return this._chapter;
  }
}
