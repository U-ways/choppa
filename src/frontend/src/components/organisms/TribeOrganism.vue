<template>
  <div class="tribe">
    <div class="row ml-0 mr-0">
      <div class="col">
        <div class="tribe__header">
          <div class="row m-0">
            <div class="col tribe__header-name text-center p-0">
              <span class="tribe__header--light">Tribe</span> {{ tribe.name }}
            </div>
          </div>
          <div class="row m-0">
            <div class="col tribe__header-underneath text-center p-0">
              {{ tribe.squads.length }} Squads
              | {{ tribe.allDistinctMembers().length }} Distinct Members
              | <router-link to="/tribe/1/edit">Edit</router-link>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="row mt-3 ml-0 mr-0">
      <div class="col-lg-6 mb-3" v-for="squad in tribe.squads" :key="squad.id">
        <SquadOrganism :squad="squad"/>
      </div>
    </div>
  </div>
</template>

<script>
import Tribe from "@/data/types/tribe";
import SquadOrganism from "@/components/organisms/SquadOrganism";

export default {
  name: "TribeOrganism",
  components: { SquadOrganism },
  props: {
    tribe: Tribe,
  },
};

</script>

<style scoped lang="scss">
@import "../../assets/scss/colours";
@import "../../assets/scss/typography";

.tribe {

  &__header {
    font-size: 1.8rem;
    font-family: $font-family-emphasis;

    &--light {
      font-weight: $font-weight-emphasis-light;
    }

    &-name {
      font-size: 1.8rem;
      font-weight: $font-weight-emphasis-medium;
      text-overflow: ellipsis;
      overflow: hidden;
      max-width: 100%;
    }

    &-underneath {
      font-family: $font-family-body;
      font-size: 0.75rem;
      font-weight: $font-weight-body-light;
    }

    &-edit {
      font-size: 1.2rem;
      line-height: 2rem;
      vertical-align: middle;
      color: red;
    }
  }
}
</style>
