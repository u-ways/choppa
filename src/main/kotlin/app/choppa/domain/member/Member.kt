package app.choppa.domain.member

import app.choppa.domain.base.BaseModel
import app.choppa.domain.chapter.Chapter
import app.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import app.choppa.domain.squad.Squad
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @Column(name = "active", columnDefinition = "BOOLEAN")
    val active: Boolean = true,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=${chapter.name})"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
