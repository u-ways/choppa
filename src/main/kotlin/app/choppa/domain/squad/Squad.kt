package app.choppa.domain.squad

import app.choppa.domain.account.Account
import app.choppa.domain.account.Account.Companion.PLACEHOLDER_ACCOUNT
import app.choppa.domain.base.BaseModel
import app.choppa.domain.member.Member
import app.choppa.domain.member.Member.Companion.NO_MEMBERS
import app.choppa.domain.tribe.Tribe
import app.choppa.domain.tribe.Tribe.Companion.PLACEHOLDER_TRIBE
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*
import javax.persistence.FetchType.EAGER

@Entity
@Table(name = "squad")
@JsonSerialize(using = SquadSerializer::class)
@JsonDeserialize(using = SquadDeserializer::class)
data class Squad(
    @Id
    @Column(name = "squad_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String,

    @Column(name = "color", columnDefinition = "INTEGER")
    val color: Int,

    // TODO(U-ways) we might be able to lazy fetch this (default is eager)
    @ManyToOne
    @JoinColumn(name = "tribe", referencedColumnName = "tribe_id")
    val tribe: Tribe,

    @ManyToMany(fetch = EAGER)
    @JoinTable(
        name = "squad_current_members",
        joinColumns = [JoinColumn(name = "squad_id", referencedColumnName = "squad_id")],
        inverseJoinColumns = [JoinColumn(name = "member_id", referencedColumnName = "member_id")]
    )
    val members: MutableList<Member> = NO_MEMBERS,

    @ManyToOne
    @JoinColumn(name = "account_id", referencedColumnName = "account_id")
    override val account: Account,
) : BaseModel {
    override fun toString() = "Squad(id=$id, name=$name, tribe=$tribe)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Squad
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_SQUADS: MutableList<Squad>
            get() = mutableListOf()
        val PLACEHOLDER_SQUAD: Squad
            get() = Squad(name = "", color = 0, tribe = PLACEHOLDER_TRIBE, account = PLACEHOLDER_ACCOUNT)
    }
}
