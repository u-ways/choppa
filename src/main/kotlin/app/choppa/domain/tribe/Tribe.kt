package app.choppa.domain.tribe

import app.choppa.domain.base.BaseModel
import app.choppa.domain.squad.Squad
import app.choppa.utils.Color.Companion.GREY
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.BatchSize
import org.hibernate.annotations.Fetch
import org.hibernate.annotations.FetchMode.SELECT
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*
import javax.persistence.FetchType.EAGER

@Entity
@Table(name = "tribe")
@JsonSerialize(using = TribeSerializer::class)
@JsonDeserialize(using = TribeDeserializer::class)
data class Tribe(
    @Id
    @Column(name = "tribe_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "TR-$id".substring(0, 15),

    @Column(name = "color", columnDefinition = "INTEGER")
    val color: Int = GREY,

    @Fetch(SELECT)
    @BatchSize(size = 10)
    @OneToMany(mappedBy = "tribe", fetch = EAGER)
    val squads: MutableList<Squad> = mutableListOf(),
) : BaseModel {
    override fun toString() = "Tribe(id=$id, name=$name)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Tribe
        if (id != other.id) return false
        return true
    }

    companion object {
        val UNASSIGNED_TRIBE = Tribe(UUID.fromString("00000000-0000-0000-0000-000000000000"), "Unassigned Squad")
    }
}
