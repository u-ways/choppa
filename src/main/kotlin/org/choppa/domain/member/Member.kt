package org.choppa.domain.member

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.choppa.domain.base.BaseModel
import org.choppa.domain.chapter.Chapter
import org.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import org.choppa.domain.history.History
import org.choppa.domain.squad.Squad
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),

    @OneToMany(mappedBy = "member")
    val history: MutableList<History> = mutableListOf()
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
