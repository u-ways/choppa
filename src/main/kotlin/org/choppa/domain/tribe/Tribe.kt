package org.choppa.domain.tribe

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.choppa.domain.base.BaseModel
import org.choppa.domain.history.History
import org.choppa.domain.squad.Squad
import org.choppa.utils.Color.Companion.GREY
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.FetchType.LAZY
import javax.persistence.Id
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "tribe")
@JsonSerialize(using = TribeSerializer::class)
@JsonDeserialize(using = TribeDeserializer::class)
data class Tribe(
    @Id
    @Column(name = "tribe_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "TR-$id".substring(0, 15),

    @Column(name = "color", columnDefinition = "INTEGER")
    val color: Int = GREY,

    @OneToMany(mappedBy = "tribe", fetch = EAGER)
    val squads: MutableList<Squad> = mutableListOf(),

    @OneToMany(mappedBy = "tribe", fetch = LAZY)
    val history: MutableList<History> = mutableListOf()
) : BaseModel {
    override fun toString() = "Tribe(id=$id, name=$name)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Tribe
        if (id != other.id) return false
        return true
    }

    companion object {
        val UNASSIGNED_TRIBE = Tribe(UUID.fromString("00000000-0000-0000-0000-000000000000"), "Unassigned Squad")
    }
}
