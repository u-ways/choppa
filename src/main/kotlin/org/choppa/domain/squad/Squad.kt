package org.choppa.domain.squad

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.choppa.domain.base.BaseModel
import org.choppa.domain.history.History
import org.choppa.domain.member.Member
import org.choppa.domain.member.Member.Companion.NO_MEMBERS
import org.choppa.domain.tribe.Tribe
import org.choppa.domain.tribe.Tribe.Companion.UNASSIGNED_TRIBE
import org.choppa.utils.Color.Companion.GREY
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.fromString
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.EAGER
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.JoinTable
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "squad")
@JsonSerialize(using = SquadSerializer::class)
@JsonDeserialize(using = SquadDeserializer::class)
data class Squad(
    @Id
    @Column(name = "squad_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "SQ-$id".substring(0, 15),

    @Column(name = "color", columnDefinition = "INTEGER")
    val color: Int = GREY,

    @ManyToOne
    @JoinColumn(name = "tribe", referencedColumnName = "tribe_id")
    val tribe: Tribe = UNASSIGNED_TRIBE,

    @ManyToMany(fetch = EAGER)
    @JoinTable(
        name = "squad_current_members",
        joinColumns = [JoinColumn(name = "squad_id", referencedColumnName = "squad_id")],
        inverseJoinColumns = [JoinColumn(name = "member_id", referencedColumnName = "member_id")]
    )
    val members: MutableList<Member> = NO_MEMBERS,

    @OneToMany(mappedBy = "squad")
    val history: MutableList<History> = mutableListOf()
) : BaseModel {
    override fun toString() = "Squad(id=$id, name=$name, tribe=$tribe)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Squad
        if (id != other.id) return false
        return true
    }

    companion object {
        val UNASSIGNED_SQUAD = Squad(fromString("00000000-0000-0000-0000-000000000000"), "Unassigned Members")
    }
}
