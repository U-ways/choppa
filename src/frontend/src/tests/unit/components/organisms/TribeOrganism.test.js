import { createLocalVue, describe, expect, it, shallowMount } from "@/tests/unit/test-imports";
import VueRouter from "vue-router";
import TribeOrganism from "@/components/organisms/TribeOrganism";
import Tribe from "@/data/types/Tribe";
import Squad from "@/data/types/Squad";
import Member from "@/data/types/Member";

const testTribe = new Tribe(1, "Test Tribe", [
  new Squad(5, "Squad One", "#FF00FF", [new Member(10, "Johnny")]),
  new Squad(6, "Squad Two", "#00FFFF", [new Member(10, "Johnny"), new Member(11, "Robert")]),
]);

describe("Tribe Organism test", () => {
  it("Renders correctly", async () => {
    const localVue = createLocalVue();
    localVue.use(VueRouter);
    const router = new VueRouter();

    const wrapper = shallowMount(TribeOrganism, {
      localVue,
      router,
      propsData: {
        tribe: testTribe,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
