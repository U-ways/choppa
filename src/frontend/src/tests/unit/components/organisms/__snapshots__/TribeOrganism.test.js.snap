// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tribe Organism test Renders correctly 1`] = `
"<div class=\\"tribe\\">
  <div class=\\"row ml-0 mr-0\\">
    <div class=\\"col\\">
      <div class=\\"tribe__header\\">
        <div class=\\"row m-0\\">
          <div class=\\"col tribe__header-name text-center p-0\\"><span class=\\"tribe__header--light\\">Tribe</span> Test Tribe
          </div>
        </div>
        <div class=\\"row m-0\\">
          <div class=\\"col tribe__header-underneath text-center p-0\\">
            2 Squads
            | 2 Distinct Members
            | <router-link-stub to=\\"/tribe/1/edit\\" tag=\\"a\\" ariacurrentvalue=\\"page\\" event=\\"click\\">Edit</router-link-stub>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class=\\"row mt-3 ml-0 mr-0\\">
    <div class=\\"col-lg-6 mb-3\\">
      <squadorganism-stub squad=\\"[object Object]\\"></squadorganism-stub>
    </div>
    <div class=\\"col-lg-6 mb-3\\">
      <squadorganism-stub squad=\\"[object Object]\\"></squadorganism-stub>
    </div>
  </div>
</div>"
`;
